pdf_langchain_chatbot.ipynb

```diff
@@ -0,0 +1,881 @@
+        "\"\"\"\n",
+        "\n",
+        "contextualize_q_prompt = ChatPromptTemplate.from_messages(\n",
+        "    [\n",
+        "        (\"system\", contextualize_q_system_prompt),\n",
+        "        MessagesPlaceholder(\"chat_history\"),\n",
+        "        (\"human\", \"{input}\")\n",
+        "    ]\n",
+        ")"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 20,
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "# Create a chain that takes conversation history and contextualizes the prompt\n",
+        "history_aware_retriever = create_history_aware_retriever(llm, ensemble_retriever, contextualize_q_prompt)"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 21,
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "# rejig qa prompt to include the chat history\n",
+        "qa_prompt = ChatPromptTemplate.from_messages(\n",
+        "    [\n",
+        "        (\"system\", system_prompt),\n",
+        "        MessagesPlaceholder(\"chat_history\"),\n",
+        "        (\"human\", \"{input}\")\n",
+        "    ]\n",
+        ")"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 22,
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "from langchain_core.chat_history import BaseChatMessageHistory\n",
+        "from langchain_community.chat_message_histories import ChatMessageHistory\n",
+        "\n",
+        "# statefully manage session history\n",
+        "store = {}\n",
+        "\n",
+        "def get_session_history(session_id: str) -> BaseChatMessageHistory:\n",
+        "    if session_id not in store:\n",
+        "        store[session_id] = ChatMessageHistory()\n",
+        "        \n",
+        "    return store[session_id]"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 23,
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "# create chains that include message history\n",
+        "qa_chain = create_stuff_documents_chain(llm, qa_prompt)\n",
+        "rag_chain = create_retrieval_chain(history_aware_retriever, qa_chain)"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 24,
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "from langchain_core.runnables.history import RunnableWithMessageHistory\n",
+        "\n",
+        "# create final chain that ties everything together\n",
+        "\n",
+        "conversation_rag_chain = RunnableWithMessageHistory(\n",
+        "    rag_chain,\n",
+        "    get_session_history,\n",
+        "    input_messages_key = \"input\",\n",
+        "    history_messages_key = \"chat_history\",\n",
+        "    output_messages_key = \"answer\"\n",
+        ")"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "## Let's try it out!"
```

Comment:
"Let's try it out on a different document!"

I'd try to clarify that this is a different usecase.